src/viam/sdk/common/proto_value.hpp

```diff
@@ -0,0 +1,324 @@
+    template <typename T>
+    friend T* dyn_cast(ProtoValue&);
+
+    /// @brief Checking cast to T, returns non-owning non-null pointer if argument is_a<T>()
+    template <typename T>
+    friend T const* dyn_cast(const ProtoValue&);
+
+    ///@}
+
+   private:
+    // This struct is our implementation of a virtual table, similar to what is created by the
+    // compiler for polymorphic types. We can't use actual polymorphic types because the
+    // implementation uses aligned stack storage, so we DIY it insted.
+    // The vtable can be thought of as defining a concept or interface which our ProtoValue types
+    // must satisfy.
+    // The first void [const]* parameter in any of the pointers below is always the `this` or `self`
+    // pointer.
+    struct vtable {
+        void (*dtor)(void*);
+        void (*copy)(void const*, void*);
+        void (*move)(void*, void*);
+        void (*to_proto)(void const*, google::protobuf::Value*);
+        int (*kind)();
+        bool (*equal_to)(void const*, void const*, const vtable&);
+    };
+
+    // Class template for a concrete model of the concept described by vtable.
+    // All construction and equality operations are implemented using defaults of the stored T.
+    template <typename T>
+    struct model {
+        model(T t) noexcept(std::is_nothrow_move_constructible<T>{});
+
+        static void dtor(void* self) noexcept;
+
+        static void copy(void const* self, void* dest);
+
+        // Clang bug prevents us from marking this noexcept but it's being called through a
+        // non-noexcept pointer anyway
+        static void move(void* self, void* dest);
+
+        static void to_proto(void const* self, google::protobuf::Value* v);
+
+        static int kind() noexcept;
+
+        static bool equal_to(void const* self, void const* other, const vtable& other_vtable);
+
+        static constexpr vtable vtable_{dtor, copy, move, to_proto, kind, equal_to};
+        T data;
+    };
+
+    // Stack-based storage for instances of model<T>.
+    // This is an RAII class which manages lifetime, pointer access, and construction of a
+    // type-erased model<T>. Note that all lifetime operations require a vtable to which the
+    // implementation defers.
+    struct storage {
+        // Local storage in an aligned_union which can hold all the possible ProtoValue types, using
+        // stand-ins for std::vector<ProtoValue> and ProtoStruct which are not available until end
+        // of class definition.
+        using BufType = std::aligned_union_t<0,
+                                             std::nullptr_t,
+                                             bool,
+                                             int,
+                                             double,
+                                             std::string,
+                                             std::vector<void*>,
+                                             std::unordered_map<std::string, void*>>;
```

Comment:
I see why you need to use `void*` as a standin. If there isn't one already, could there be a static assertion later that the sizes are the same whether or not it is `void*` or `ProtoStruct`?

Reply:
Actually, even better, add some static assertions at .cpp scope once you have access to the full `ProtoValue` type.

Reply:
There is the following in the cpp file, maybe this is adequate?
https://github.com/viamrobotics/viam-cpp-sdk/blob/b659c767e9d3d250692bf3a2a50598e78266d252/src/viam/sdk/common/proto_value.cpp#L146-L148

I think that more accurately names the object of concern here, ie, 
- is the storage large enough for any type being constructed?
- is there any type being constructed with alignment too strict for the storage alignment?

We could assert on the `sizeof(ProtoStruct)` too though, what do you think?

Reply:
I think it is worth explicitly validating that `alignof` and `sizeof` for `std::vector<void*>` and `std::unordered_map<std::string, void*>` match the `alignof` and `sizeof` for those some types with `ProtoStruct` as the mapped type.